ui/raidboss/emulator/data/PopupTextAnalysis.js

```diff
@@ -1,20 +1,68 @@
 import EmulatorCommon from '../EmulatorCommon';
 import StubbedPopupText from '../overrides/StubbedPopupText';
 
-export default class PopupTextAnalysis extends StubbedPopupText {
-  OnTriggerInternal(trigger, matches, currentTime) {
-    this.currentTriggerStatus = {
-      initialData: EmulatorCommon.cloneData(this.data),
-      condition: undefined,
-      response: false,
-      result: false,
-      delay: false,
-      suppressed: false,
-      executed: false,
-      promise: undefined,
+class Resolver {
+  constructor() {
+    this.promise = null;
+    this.run = null;
+    this.delayTo = null;
+  }
+  async IsResolved(log) {
+    if (this.delayTo) {
+      if (this.delayTo < log.timestamp) {
+        this.delayTo = null;
+        this.delay();
+        await this.delayPromise;
+      } else {
+        return false;
+      }
+    }
+    if (!this.delayTo) {
+      if (this.promise)
+        await this.promise();
+      if (this.run)
+        await this.run();
+      if (this.final)
+        this.final();
+      return true;
+    }
+    return false;
+  }
+  setDelay(delayTo) {
+    this.delayTo = delayTo;
+    return this.delayPromise = new Promise((res) => {
+      this.delay = res;
+    });
+  }
+  setPromise(promise) {
+    this.promise = async () => {
+      await promise;
```

Comment:
Can this just be this.promise = promise?

Reply:
Adjusted for next commit. This used to do more in the inner async function.